src/app/clusters/commodity-price-server/commodity-price-server.cpp

```diff
@@ -388,46 +371,128 @@ CHIP_ERROR Instance::SetCurrentPrice(const DataModel::Nullable<Structs::Commodit
     // Check PeriodStart  < PeriodEnd (if not null)
     if (!newValue.IsNull())
     {
-        Structs::CommodityPriceStruct::Type tempValue;
         if (!newValue.Value().periodEnd.IsNull() && (newValue.Value().periodStart > newValue.Value().periodEnd.Value()))
         {
             return CHIP_IM_GLOBAL_STATUS(ConstraintError);
         }
+
+        if (!newValue.Value().price.HasValue() && !newValue.Value().priceLevel.HasValue())
+        {
+            // Must have Price or PriceLevel
+            return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+        }
     }
 
-    mCurrentPrice = newValue;
+    // Do a deep copy of the newValue into mCurrentPrice
+    CopyPrice(newValue);
 
-    ChipLogDetail(AppServer, "mCurrentPrice updated");
+    ChipLogDetail(AppServer, "Endpoint: %d - mCurrentPrice updated", mEndpointId);
     MatterReportingAttributeChangeCallback(mEndpointId, CommodityPrice::Id, CurrentPrice::Id);
 
     // generate a PriceChange Event
-    SendPriceChangeEvent();
+    GeneratePriceChangeEvent();
 
     return CHIP_NO_ERROR;
 }
 
-Status Instance::SendPriceChangeEvent()
+CHIP_ERROR Instance::CopyStringToSpan(const char * src, Platform::ScopedMemoryBuffer<char> & bufferOut, CharSpan & spanOut)
 {
+    size_t len = strlen(src);
+    if (bufferOut.Get() != nullptr)
+    {
+        bufferOut.Free();
+    }
 
-    CHIP_ERROR err;
-    DataModel::Nullable<Structs::CommodityPriceStruct::Type> priceStruct;
+    if (!bufferOut.Calloc(len))
+    {
+        return CHIP_ERROR_NO_MEMORY;
```

Comment:
But callers are not checking the return value, so they will happily point their spans to garbage....